File: companion/backend/feishu.py
"""FeishuBackend: 通过飞书 Bot 与真人私聊的桥接

当前是骨架实现。完整闭环需要:
1. 自建飞书应用, 拿到 app_id / app_secret。
2. 配置事件订阅 URL, 接收 im.message.receive_v1。
3. 本地跑一个 webhook 服务接事件 → 写入下面的历史存储 → UI 拉取。
4. `send` 调 /open-apis/im/v1/messages 以 bot 身份发消息。

骨架部分仅把输入写本地 SQLite, 便于先把 UI 切换接通; 真正的 API 调用和
webhook 事件接入单独实现后再替换。
"""

from __future__ import annotations

import os
import sqlite3
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import AsyncIterator

import urllib.error
import urllib.request
import json

from .base import BackendMetadata, ChatBackend, Message


TENANT_TOKEN_ENDPOINT = "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal"
SEND_MESSAGE_ENDPOINT = "https://open.feishu.cn/open-apis/im/v1/messages"


class FeishuBackend(ChatBackend):
    def __init__(
        self,
        contact_slug: str,
        *,
        chat_id: str,
        display_name: str,
        storage_dir: str | Path,
        app_id: str | None = None,
        app_secret: str | None = None,
    ):
        self._slug = contact_slug
        self._chat_id = chat_id
        self._name = display_name
        self._app_id = app_id or os.environ.get("FEISHU_APP_ID")
        self._app_secret = app_secret or os.environ.get("FEISHU_APP_SECRET")

        self._storage = Path(storage_dir)
        self._storage.mkdir(parents=True, exist_ok=True)
        self._db_path = self._storage / f"feishu-{contact_slug}.sqlite"
        self._init_db()

        self._token_cache: tuple[str, float] | None = None

    # ----- metadata -----

    @property
    def metadata(self) -> BackendMetadata:
        has_creds = bool(self._app_id and self._app_secret)
        notice = (
            "真人 · 通过飞书 Bot 中转。对方看到的是机器人身份。"
            if has_creds
            else "⚠️ 未配置 FEISHU_APP_ID / FEISHU_APP_SECRET, 当前只能看历史, 发消息会入列但不会真发。"
        )
        return BackendMetadata(
            source="feishu",
            display_name=self._name,
            persona_slug=None,
            feishu_chat_id=self._chat_id,
            read_only=not has_creds,
            notice=notice,
        )

    # ----- history -----

    def _init_db(self) -> None:
        with sqlite3.connect(self._db_path) as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS messages (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    direction TEXT NOT NULL CHECK(direction IN ('in', 'out', 'pending')),
                    role TEXT NOT NULL,
                    content TEXT NOT NULL,
                    ts TEXT NOT NULL,
                    feishu_message_id TEXT
                )
                """
            )

    async def history(self, limit: int = 50) -> list[Message]:
        with sqlite3.connect(self._db_path) as conn:
            rows = conn.execute(
                "SELECT role, content, ts FROM messages ORDER BY id DESC LIMIT ?",
                (limit,),
            ).fetchall()
        rows.reverse()
        return [Message(role=r[0], content=r[1], timestamp=r[2], source="feishu") for r in rows]

    def record_incoming(self, content: str, feishu_message_id: str | None = None) -> None:
        """Webhook 接收到对方消息时由事件处理器调用."""
        ts = datetime.now(tz=timezone.utc).astimezone().isoformat()
        with sqlite3.connect(self._db_path) as conn:
            conn.execute(
                "INSERT INTO messages(direction, role, content, ts, feishu_message_id) VALUES(?,?,?,?,?)",
                ("in", "assistant", content, ts, feishu_message_id),
            )

    # ----- send -----

    async def send(self, text: str) -> AsyncIterator[str]:
        ts = datetime.now(tz=timezone.utc).astimezone().isoformat()
        direction = "out" if (self._app_id and self._app_secret) else "pending"

        with sqlite3.connect(self._db_path) as conn:
            conn.execute(
                "INSERT INTO messages(direction, role, content, ts) VALUES(?,?,?,?)",
                (direction, "user", text, ts),
            )

        if direction == "pending":
            async def _skipped() -> AsyncIterator[str]:
                yield f"[未配置飞书凭证, 消息已入列但未发出: {text[:40]}]"

            return _skipped()

        try:
            feishu_message_id = self._send_to_feishu(text)

            async def _ok() -> AsyncIterator[str]:
                yield f"[已发送 · message_id={feishu_message_id}]"
                yield "\n对方回复会通过事件订阅推回, 请保持 webhook 服务运行。"

            return _ok()
        except Exception as exc:  # noqa: BLE001
            async def _fail() -> AsyncIterator[str]:
                yield f"[发送失败: {exc}]"

            return _fail()

    # ----- 飞书 API ----

    def _tenant_token(self) -> str:
        if self._token_cache and self._token_cache[1] > time.time() + 60:
            return self._token_cache[0]
        assert self._app_id and self._app_secret
        body = json.dumps({"app_id": self._app_id, "app_secret": self._app_secret}).encode("utf-8")
        req = urllib.request.Request(
            TENANT_TOKEN_ENDPOINT,
            data=body,
            headers={"Content-Type": "application/json"},
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read())
        if data.get("code") != 0:
            raise RuntimeError(f"获取 tenant_access_token 失败: {data}")
        token = data["tenant_access_token"]
        expire = time.time() + int(data.get("expire", 1800))
        self._token_cache = (token, expire)
        return token

    def _send_to_feishu(self, text: str) -> str:
        token = self._tenant_token()
        payload = {
            "receive_id": self._chat_id,
            "msg_type": "text",
            "content": json.dumps({"text": text}, ensure_ascii=False),
        }
        body = json.dumps(payload).encode("utf-8")
        url = f"{SEND_MESSAGE_ENDPOINT}?receive_id_type=chat_id"
        req = urllib.request.Request(
            url,
            data=body,
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
            },
        )
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = json.loads(resp.read())
        except urllib.error.HTTPError as exc:
            raise RuntimeError(f"HTTP {exc.code}: {exc.read().decode('utf-8', errors='ignore')}") from exc
        if data.get("code") != 0:
            raise RuntimeError(f"飞书 send_message 失败: {data}")
        msg_id = data.get("data", {}).get("message_id", "")
        with sqlite3.connect(self._db_path) as conn:
            conn.execute(
                "UPDATE messages SET feishu_message_id = ? WHERE id = (SELECT MAX(id) FROM messages)",
                (msg_id,),
            )
        return msg_id

File: packages/companion/src/backend/feishu.ts
import type { BackendMetadata, Message } from "@petchat/shared";
import type { ChatBackend } from "./base.js";
import { openDb } from "./sqlite.js";
import path from "node:path";

const TENANT_TOKEN_ENDPOINT =
  "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal";
const SEND_MESSAGE_ENDPOINT =
  "https://open.feishu.cn/open-apis/im/v1/messages";

export interface FeishuBackendOptions {
  contactSlug: string;
  chatId: string;
  displayName: string;
  storageDir: string;
  appId?: string;
  appSecret?: string;
}

export class FeishuBackend implements ChatBackend {
  readonly metadata: BackendMetadata;
  private db: ReturnType<typeof openDb>;
  private appId?: string;
  private appSecret?: string;
  private chatId: string;
  private tokenCache: { token: string; expiresAt: number } | null = null;

  constructor(opts: FeishuBackendOptions) {
    this.appId = opts.appId ?? process.env.FEISHU_APP_ID;
    this.appSecret = opts.appSecret ?? process.env.FEISHU_APP_SECRET;
    this.chatId = opts.chatId;

    const hasCreds = !!(this.appId && this.appSecret);
    this.metadata = {
      source: "feishu",
      displayName: opts.displayName,
      feishuChatId: opts.chatId,
      readOnly: !hasCreds,
      notice: hasCreds
        ? "真人 · 通过飞书 Bot 中转。对方看到的是机器人身份。"
        : "⚠️ 未配置 FEISHU_APP_ID / FEISHU_APP_SECRET, 当前只能看历史, 发消息会入列但不会真发。",
    };

    const dbPath = path.join(opts.storageDir, `feishu-${opts.contactSlug}.sqlite`);
    this.db = openDb(dbPath);
    this.db.exec(
      `CREATE TABLE IF NOT EXISTS messages (
         id INTEGER PRIMARY KEY AUTOINCREMENT,
         direction TEXT NOT NULL CHECK(direction IN ('in','out','pending')),
         role TEXT NOT NULL,
         content TEXT NOT NULL,
         ts TEXT NOT NULL,
         feishu_message_id TEXT
       )`,
    );
  }

  async history(limit = 50): Promise<Message[]> {
    const rows = this.db
      .prepare("SELECT role, content, ts FROM messages ORDER BY id DESC LIMIT ?")
      .all(limit) as { role: string; content: string; ts: string }[];
    return rows.reverse().map((r) => ({
      role: r.role as Message["role"],
      content: r.content,
      timestamp: r.ts,
      source: "feishu" as const,
    }));
  }

  /** Called by the webhook receiver when ta replies. */
  recordIncoming(content: string, feishuMessageId?: string): void {
    const ts = new Date().toISOString();
    this.db
      .prepare(
        "INSERT INTO messages(direction, role, content, ts, feishu_message_id) VALUES(?,?,?,?,?)",
      )
      .run("in", "assistant", content, ts, feishuMessageId ?? null);
  }

  async *send(text: string): AsyncIterable<string> {
    const ts = new Date().toISOString();
    const direction = this.appId && this.appSecret ? "out" : "pending";
    this.db
      .prepare(
        "INSERT INTO messages(direction, role, content, ts) VALUES(?,?,?,?)",
      )
      .run(direction, "user", text, ts);

    if (direction === "pending") {
      yield `[未配置飞书凭证, 消息已入列但未发出: ${text.slice(0, 40)}]`;
      return;
    }

    try {
      const messageId = await this.sendToFeishu(text);
      yield `[已发送 · message_id=${messageId}]`;
      yield "\n对方回复会通过事件订阅推回, 请保持 webhook 服务运行。";
    } catch (err) {
      yield `[发送失败: ${(err as Error).message}]`;
    }
  }

  // ---------- Feishu API ----------

  private async tenantToken(): Promise<string> {
    if (this.tokenCache && this.tokenCache.expiresAt > Date.now() + 60_000) {
      return this.tokenCache.token;
    }
    const resp = await fetch(TENANT_TOKEN_ENDPOINT, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ app_id: this.appId, app_secret: this.appSecret }),
    });
    const data = (await resp.json()) as {
      code: number;
      tenant_access_token?: string;
      expire?: number;
    };
    if (data.code !== 0 || !data.tenant_access_token) {
      throw new Error(`获取 tenant_access_token 失败: ${JSON.stringify(data)}`);
    }
    this.tokenCache = {
      token: data.tenant_access_token,
      expiresAt: Date.now() + (data.expire ?? 1800) * 1000,
    };
    return data.tenant_access_token;
  }

  private async sendToFeishu(text: string): Promise<string> {
    const token = await this.tenantToken();
    const url = `${SEND_MESSAGE_ENDPOINT}?receive_id_type=chat_id`;
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        receive_id: this.chatId,
        msg_type: "text",
        content: JSON.stringify({ text }),
      }),
    });
    const data = (await resp.json()) as {
      code: number;
      data?: { message_id?: string };
      msg?: string;
    };
    if (data.code !== 0) {
      throw new Error(`飞书 send_message 失败: ${JSON.stringify(data)}`);
    }
    const msgId = data.data?.message_id ?? "";
    this.db
      .prepare(
        "UPDATE messages SET feishu_message_id = ? WHERE id = (SELECT MAX(id) FROM messages)",
      )
      .run(msgId);
    return msgId;
  }
}
